RealmSwift/Geospatial.swift

```diff
@@ -0,0 +1,201 @@
+
+ - warning: This class cannot be persisted and can only be use within a geospatial `geoWithin` query.
+ */
+public typealias GeoBox = RLMGeospatialBox
+
+public extension GeoBox {
+    /// Initialize a `GeoBox`, with values for bottom left corner and top right corner.
+    ///
+    /// - Parameter bottomLeft: The bottom left corner of the rectangle.
+    /// - Parameter topRight: The top right corner of the rectangle.
+    convenience init?(bottomLeft: (Double, Double), topRight: (Double, Double)) {
+        guard let bottomLeftPoint = GeoPoint(bottomLeft.0, bottomLeft.1),
+              let topRightPoint = GeoPoint(topRight.0, topRight.1) else {
+            return nil
+        }
+        self.init(bottomLeft: bottomLeftPoint, topRight: topRightPoint)
+    }
+}
+
+/**
+ A class that represents a polygon, that can be used in a geospatial `geoWithin`query.
+
+ A `GeoPolygon` describes a shape conformed of and outer `Polygon`, called `outerRing`,
+ and 0 or more inner `Polygon`s, called `holes`, which represents an unlimited number of internal holes
+ inside the outer `Polygon`.
+ A `Polygon` describes a shape conformed by at least three segments, where the last and the first `GeoPoint`
+ must be the same to indicate a closed polygon (meaning you need at least 4 points to define a polygon).
+ Inner holes in a `GeoPolygon` must  be entirely inside the outer ring
+ 
+ A `hole` has the following restrictions:
+ - Holes may not cross, i.e. the boundary of a hole may not intersect both the interior and the exterior of any other
+   hole.
+ - Holes may not share edges, i.e. if a hole contains and edge AB, the no other hole may contain it.
+ - Holes may share vertices, however no vertex may appear twice in a single hole.
+ - No hole may be empty.
+ - Only one nesting is allowed.
+
+ - warning: This class cannot be persisted and can only be use within a geospatial `geoWithin` query.
+ */
+public typealias GeoPolygon = RLMGeospatialPolygon
+
+public extension GeoPolygon {
+    /// Initialize a `GeoPolygon`, with values for bottom left corner and top right corner.
+    ///
+    /// Returns `nil` if the `GeoPoints` representing a polygon (outer ring or holes), don't have at least 4 points.
+    /// Returns `nil` if the first and the last `GeoPoint` in a polygon are not the same.
+    ///
+    /// - Parameter outerRing: The polygon's external (outer) ring.
+    /// - Parameter holes: The holes (if any) in the polygon.
+    convenience init?(outerRing: [(Double, Double)], holes: [[(Double, Double)]]? = nil) {
+        let outerRingPoints = outerRing.map { GeoPoint(latitude: $0.0, longitude: $0.1) }
+        guard !outerRingPoints.contains(nil) else {
+            return nil
+        }
+        var holesPoints: [[GeoPoint]]?
+        if let holes = holes {
+            let holesMapped = holes.map { $0.map { GeoPoint(latitude: $0.0, longitude: $0.1) } }
+            guard !holesMapped.flatMap({ $0 }).contains(nil) else {
```

Comment:
This is a really expensive way of writing this check; it's producing a large temporary array only to immediately discard it after checking if it contains nil. One way to avoid this and the duplicated conversion would be `holesPoints = holes.map { $0.compactMap(GeoPoint.init) }; guard zip(holes, holesPoints).allSatisfy({ $0.count == $1.count }) else { return nil }`.